File: src/main/java/ru/octol1ttle/flightassistant/util/events/FireworkBoostCallback.java
package ru.octol1ttle.flightassistant.util.events;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.projectile.FireworkRocketEntity;

@FunctionalInterface
public interface FireworkBoostCallback {
    Event<FireworkBoostCallback> EVENT = EventFactory.createArrayBacked(
            FireworkBoostCallback.class,
            (listeners) -> (rocket, shooter) -> {
                for (FireworkBoostCallback event : listeners) {
                    event.onFireworkBoost(rocket, shooter);
                }
            }
    );

    void onFireworkBoost(FireworkRocketEntity rocket, LivingEntity shooter);
}

File: src/client/kotlin/ru/octol1ttle/flightassistant/api/event/FireworkBoostCallback.kt
package ru.octol1ttle.flightassistant.api.event

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.network.ClientPlayerEntity
import net.minecraft.entity.projectile.FireworkRocketEntity


fun interface FireworkBoostCallback {
    fun onFireworkBoost(rocket: FireworkRocketEntity?, shooter: ClientPlayerEntity)

    companion object {
        @JvmField
        val EVENT: Event<FireworkBoostCallback> =
            EventFactory.createArrayBacked(FireworkBoostCallback::class.java)
            { listeners: Array<FireworkBoostCallback> ->
                FireworkBoostCallback { rocket: FireworkRocketEntity?, shooter: ClientPlayerEntity ->
                    for (event in listeners) {
                        event.onFireworkBoost(rocket, shooter)
                    }
                }
            }
    }
}
